File: CodeGen/DuckTyping.cs
using System;
using System.Collections.Generic;
using System.Diagnostics.Contracts;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;

namespace Mvvm.CodeGen
{
    public static class DuckTyping
    {
        static MemberInfo TryFindMember(Type thingType, MemberInfo duckMember)
        {
            return null;
        }

        public static bool CanCast<T>(object thing)
        {
            Contract.Requires(thing != null);
            Contract.Requires(typeof(T).IsInterface);

            var duckType = typeof(T);
            var thingType = thing.GetType();
            var members = FindAllMembers(duckType);
            Contract.Assume(members.All(m => m is MethodInfo || m is PropertyInfo));
            var thingMembers = members.Select(m => TryFindMember(thingType, m));

            return true;
        }

        private static IEnumerable<MemberInfo> FindAllMembers(Type duckType)
        {
            var interfaces = duckType.Flatten(t => t.GetInterfaces()).Distinct();
            var members = from i in interfaces
                          from ii in i.GetMembers()
                          select ii;
            return members.Distinct();
        }

        public static T Cast<T>(object thing)
        {
            Contract.Requires(CanCast<T>(thing));
            return default(T);
        }
    }
}

File: DuckTypingFS/DuckTyping.fs
namespace Mvvm.Codegen

open System
open System.Reflection
open System.Reflection.Emit
open System.Diagnostics.Contracts
open System.Collections
open System.Collections.Generic
open System.Threading
open System.Threading.Tasks


type MappingKey = { TDuck : Type; TThing : Type }
type MappingValue = { T : Type; Init : Action<Object,Object> }

type DuckTyping =
  class
    static member private _mappings = new Dictionary<MappingKey, MappingValue>()
    static member private _lock = new Object()

    ///Maps a Thing-Type to an interface-Type.
    ///Returns the mapped type and a function to initialise it
    static member Map tDuck tThing =
        let key = { TDuck = tDuck; TThing = tThing }
        let mapping = lock (DuckTyping._lock) (fun () ->
            if DuckTyping._mappings.ContainsKey (key) then
                DuckTyping._mappings.[key]
            else
                let (t,init) = DuckTypingFSInternal.mapType tDuck tThing
                let value = { T = t; Init = init }
                DuckTyping._mappings.Add(key, value)
                value  )
        mapping

    ///Casts a Thing to an interface IDuck.
    ///The thing must implement all members of IDuck.
    static member Cast<'TDuck when 'TDuck: null and 'TDuck : not struct> thing =
        Contract.Requires (thing <> null)
        let tDuck = typedefof<'TDuck>
        let tThing = thing.GetType()
        
        let mapping = DuckTyping.Map tDuck tThing

        let instance = Activator.CreateInstance(mapping.T, [| thing |]) :?> 'TDuck
        mapping.Init.Invoke (instance,thing)
        instance
  end
